src/lua/datetime.lua

```diff
@@ -172,11 +256,69 @@ local function check_ymd_range(y, M, d)
         return
     end
 ::min_err::
-    error(('date %d-%02d-%02d is less than minimum allowed %d-%02d-%02d'):
-        format(y, M, d, MIN_DATE_YEAR, MIN_DATE_MONTH, MIN_DATE_DAY))
+    error(('date %d-%02d-%02d is less than minimum allowed %s'):
+        format(y, M, d, MIN_DATE_TEXT))
 ::max_err::
-    error(('date %d-%02d-%02d is greater than maximum allowed %d-%02d-%02d'):
-        format(y, M, d, MAX_DATE_YEAR, MAX_DATE_MONTH, MAX_DATE_DAY))
+    error(('date %d-%02d-%02d is greater than maximum allowed %s'):
+        format(y, M, d, MAX_DATE_TEXT))
+end
+
+local function bool2int(b)
+    return b and 1 or 0
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 or nsec >= NANOS_PER_SEC then
+        secs = secs + math_floor(nsec / NANOS_PER_SEC)
+        nsec = nsec % NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function interval_decouple_args(obj)
```

Comment:
Yup, now we handle it. Thanks for the catch!:

```
tarantool> date.interval.new{day=1e21}
---
- error: value 1e+21 of day is out of allowed range [-4295055470.25, 4295055470.25]
...

tarantool> date.interval.new{week=1e21}
---
- error: value 1e+21 of week is out of allowed range [-613579352.89286, 613579352.89286]
...

tarantool> date.interval.new{hour=1e21}
---
- error: value 1e+21 of hour is out of allowed range [-103081331286, 103081331286]
...

tarantool> date.interval.new{min=1e21}
---
- error: value 1e+21 of min is out of allowed range [-6184879877160, 6184879877160]
...

tarantool> date.interval.new{sec=1e21}
---
- error: value 1e+21 of sec is out of allowed range [-3.710927926296e+14, 3.710927926296e+14]
...
```
Though, I've left nanoseconds range check out of touch, because I do `normalize_nsec` in each case, so it get normalized according to current rules (at the moment normalized by modulo 1e9, in a future for leap seconds it might get normalized by modulo 2e9).

[This lack of checks for nanoseconds is debatable, and might be changed]

The tricky part for all those checks for each involved attributed - is to extensively to cover these cases by tests

Reply:
```diff
@@ -105,6 +105,7 @@ local MAX_DATE_DAY = 11
 -- about 1 day in 128 years.
 local AVERAGE_DAYS_YEAR = 365.25
 local AVERAGE_DAYS_MONTH = (AVERAGE_DAYS_YEAR / 12)
+local AVERAGE_WEEK_YEAR = (AVERAGE_DAYS_YEAR / 7)
 local INT_MAX = 2147483647
 local INT_MIN = -2147483648
 -- -5879610-06-22
@@ -113,6 +114,13 @@ local MIN_DATE_TEXT = ('%d-%02d-%02d'):format(MIN_DATE_YEAR, MIN_DATE_MONTH,
 -- 5879611-07-11
 local MAX_DATE_TEXT = ('%d-%02d-%02d'):format(MAX_DATE_YEAR, MAX_DATE_MONTH,
                                               MAX_DATE_DAY)
+local MAX_YEAR_RANGE = (MAX_DATE_YEAR - MIN_DATE_YEAR)
+local MAX_MONTH_RANGE = (MAX_YEAR_RANGE * 12)
+local MAX_WEEK_RANGE = (MAX_YEAR_RANGE * AVERAGE_WEEK_YEAR)
+local MAX_DAY_RANGE = (MAX_YEAR_RANGE * AVERAGE_DAYS_YEAR)
+local MAX_HOUR_RANGE = (MAX_DAY_RANGE * 24)
+local MAX_MIN_RANGE = (MAX_HOUR_RANGE * 60)
+local MAX_SEC_RANGE = (MAX_DAY_RANGE * SECS_PER_DAY)
 
 local date_tostr_stash =
     buffer.ffi_stash_new(string.format('char[%s]', TOSTRING_BUFSIZE))
@@ -281,6 +289,24 @@ local function check_ymd_range(y, M, d)
         format(y, M, d, MAX_DATE_TEXT))
 end
 
+-- check v value against maximum/minimum possible values
+-- if there is nil then simply return default 0
+local function checked_max_value(v, max, txt)
+    if v == nil then
+        return 0
+    end
+    if type(v) ~= 'number' then
+        error(('numeric value expected, but received %s'):
+              format(type(v)), 2)
+    end
+    if v > -max and v < max then
+        return v
+    end
+    error(('value %s of %s is out of allowed range [%s, %s]'):
+            format(v, txt, -max, max), 4)
+end
+
+
 local function bool2int(b)
     return b and 1 or 0
 end
@@ -297,15 +323,16 @@ local function interval_decouple_args(obj)
     if is_interval(obj) then
         return obj.year, obj.month, obj.sec, obj.nsec
     end
-    local year = obj.year or 0
-    local month = obj.month or 0
+    local year = checked_max_value(obj.year, MAX_YEAR_RANGE, 'year')
+    local month = checked_max_value(obj.month, MAX_MONTH_RANGE, 'month')
 
     local secs = 0
-    secs = secs + (7 * SECS_PER_DAY) * (obj.week or 0)
-    secs = secs + SECS_PER_DAY * (obj.day or 0)
-    secs = secs + (60 * 60) * (obj.hour or 0)
-    secs = secs + 60 * (obj.min or 0)
-    local sec = obj.sec
+    secs = secs + (7 * SECS_PER_DAY) *
+                  checked_max_value(obj.week, MAX_WEEK_RANGE, 'week')
+    secs = secs + SECS_PER_DAY * checked_max_value(obj.day, MAX_DAY_RANGE, 'day')
+    secs = secs + (60 * 60) * checked_max_value(obj.hour, MAX_HOUR_RANGE, 'hour')
+    secs = secs + 60 * checked_max_value(obj.min, MAX_MIN_RANGE, 'min')
+    local sec = checked_max_value(obj.sec, MAX_SEC_RANGE, 'sec')
     local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
     local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
                        bool2int(msec ~= nil)
```